### PR TITLE
Optionally parse list of tags to filter indexing by

### PR DESCRIPTION
**Pull request by karenzshea**
Closes https://github.com/mapbox/route-annotator/issues/24

First pass implementation of filtering on OSM file by certain tags.

Remaining tasks
- [x] Figure out segfault
- [x] Write tests
- [x] Fix TODOs around optional handling
- [x] Figure out handling when no ways are returned
- [x] Allow `loadOSMExtract` to accept an array of files as well as one file string

### CODE REVIEW

File: src/extractor.cpp [hunk 1]
```diff
@@ -30,6 +31,55 @@ typedef osmium::index::map::SparseMemArray<osmium::unsigned_object_id_type, osmi
...
 typedef osmium::handler::NodeLocationsForWays<index_pos_type, index_neg_type> location_handler_type;
 
+void Extractor::ParseTags(std::ifstream &tagfile)
+{
+    std::string line;
+    while (std::getline(tagfile, line))
+    {
+        tags_filter.add_rule(true, osmium::TagMatcher(line));
+    }
+    tagfile.close();
```
Comment (daniel-j-h) [inline src/extractor.cpp]: Remove; the ifstream gets closed automatically

File: src/extractor.cpp [hunk 2]
```diff
@@ -30,6 +31,55 @@ typedef osmium::index::map::SparseMemArray<osmium::unsigned_object_id_type, osmi
...
+    if (!tagfilename.empty())
+    {
+        std::cerr << "Parsing " << tagfilename << " ... " << std::flush;
+        std::ifstream tagfile(tagfilename);
+        if (!tagfile.is_open())
+        {
+            throw std::runtime_error(strerror(errno));
+        }
+        ParseTags(tagfile);
+    }
```
Comment (daniel-j-h) [inline src/extractor.cpp]: This ending scope closes the ifstream (in its destructor)

File: src/extractor.cpp [hunk 3]
```diff
@@ -102,20 +152,37 @@ void Extractor::way(const osmium::Way &way)
...
+             std::strcmp(highway, "trunk") == 0 || std::strcmp(highway, "trunk_link") == 0 ||
+             std::strcmp(highway, "primary") == 0 || std::strcmp(highway, "primary_link") == 0 ||
+             std::strcmp(highway, "secondary") == 0 ||
+             std::strcmp(highway, "secondary_link") == 0 || std::strcmp(highway, "tertiary") == 0 ||
+             std::strcmp(highway, "tertiary_link") == 0 ||
+             std::strcmp(highway, "residential") == 0 ||
+             std::strcmp(highway, "living_street") == 0 ||
+             std::strcmp(highway, "unclassified") == 0 || std::strcmp(highway, "service") == 0 ||
+             std::strcmp(highway, "ferry") == 0 || std::strcmp(highway, "movable") == 0 ||
+             std::strcmp(highway, "shuttle_train") == 0 || std::strcmp(highway, "default") == 0);
```
Comment (daniel-j-h) [inline src/extractor.cpp]: What about creating a function taking a vector of tags and checking all of them?

Or better put all tags here into a hashset and do constant time `hset.count(highway) > 1` checks

File: src/extractor.hpp [hunk 4]
```diff
@@ -39,10 +43,13 @@ struct Extractor final : osmium::handler::Handler
      *
      * @param way the current way being processed
      */
+    bool FilterWay(const osmium::Way &way);
     void way(const osmium::Way &way);
+    void ParseTags(std::ifstream &tagfile);
```
Comment (danpat) [inline src/extractor.hpp]: These two new methods might be better off being private (or at least protected).  Not sure we'd ever want to call them from outside the class constructor, they're mostly just internal helpers.

File: src/extractor.cpp [hunk 5]
```diff
@@ -30,6 +31,55 @@ typedef osmium::index::map::SparseMemArray<osmium::unsigned_object_id_type, osmi
 // index_pos_type;
 typedef osmium::handler::NodeLocationsForWays<index_pos_type, index_neg_type> location_handler_type;
 
+void Extractor::ParseTags(std::ifstream &tagfile)
+{
+    std::string line;
+    while (std::getline(tagfile, line))
+    {
+        tags_filter.add_rule(true, osmium::TagMatcher(line));
+        std::cout << "tag added: " << line << std::endl;
```
[the remark this answers is not in the document]
Comment (karenzshea) [reply src/extractor.cpp]: Ok, does it make sense then to switch to using `cout` for info type log lines and `cerr` for error specific messages?

[the remark this answers is not in the document]
Comment (danpat) [reply src/extractor.cpp]: This is a good writeup on the philosophy of `stderr` vs `stdout`:
https://www.jstorimer.com/blogs/workingwithcode/7766119-when-to-use-stderr-instead-of-stdout

TL;DR - normal messages from the program, and things explicitly requested by adding parameters should go to `stdout`, out-of-the-ordinary messages and errors should go to `stderr`.

**Comment by danpat** [comment]
@karenzshea I've rebased this on `master`, which has pulled in the changes from https://github.com/mapbox/route-annotator/pull/30.

I've tagged and published 0.1.0-rc3 from this branch.

There's only one outstanding question: should all tags from a matching way be indexed, or only the tags indicated by the filter?

Currently, `master` separates the "tags to use to filter ways" from the "tags to store in the index".  I've duplicated that logic on this branch, but it caused one test to fail:

https://github.com/mapbox/route-annotator/pull/25/files#diff-879eb6ab79a85f13ac84c620132435e1R100

If you're OK with this behaviour (only index the tags that we mention in the filter), then I think this PR is ready to merge to `master` and do a release of 0.1.0 on.

If you think we should index all tags on matching ways, then this `if` statement:

https://github.com/mapbox/route-annotator/pull/25/files#diff-3cb06dc074df26ca8f412b291b6902ccR171

should be removed, and we should index all tags.  For the 0.1.0 release, I don't think it matters if we only index tags listed by the filter, we can certainly make this more flexible in a future release.

**Comment by karenzshea** [comment]
@danpat I'm OK with the behavior of only storing tags that are in the filter. I only wrote the test that way to check that the way I anticipated was being detected.

Thanks for rebasing this after the other PR was merged!
